File: Java version/app/src/main/java/it/unical/mat/embasp/languages/Mapper.java
package it.unical.mat.embasp.languages;

import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import it.unical.mat.embasp.languages.asp.IllegalTermException;

/**
 * Base class
 * Contains methods used to transform Objects into {@link it.unical.mat.embasp.base.InputProgram}
 */

public abstract class Mapper {

	protected final Map<String, Class<?>> predicateClass = new HashMap<>();

	protected final Map<Class<?>, Map<String, Method>> classSetterMethod = new HashMap<>();

	protected abstract String getActualString(String predicate, HashMap<Integer, Object> parametersMap) throws IllegalTermException;

	public Class<?> getClass(final String predicate) {
		return predicateClass.get(predicate);
	}
	
	/**
	 * Returns a Set of Objects for the given string representing a list of atoms
	 *
	 * @param string
	 *            String from witch data are extrapolated
	 * @return Set of Objects for the given String data
	 * @throws IllegalAccessException, InstantiationException, InvocationTargetException
	 */
	public Set<Object> getObjects(final String atomsList) throws IllegalAccessException, InstantiationException, InvocationTargetException {
		buildParseTree(atomsList);
		
		final Set <Object> objects = new HashSet <> ();
		String predicate = getId();
		
		while(predicate != null) {
			final Class<?> cl = getClass(predicate);
			final String[] parameters = getParam();
			predicate = getId();
			
			if (cl == null || parameters == null)
				continue;
			
			final Object obj = cl.newInstance();

			populateObject(cl, parameters, obj);
			objects.add(obj);
		}

		return objects;
	}
	
	/**
	 * @param string
	 *            The string representing a list of atoms received as parameter of getObjects
	 */
	protected abstract void buildParseTree(final String atomsList);
	
	/**
	 * @return The predicate name
	 */
	protected abstract String getId();
	
	/**
	 * @return All the Terms
	 */
	protected abstract String[] getParam();
	
	/**
	 * Returns an Object for the given string
	 *
	 * @param string
	 *            String from witch data are extrapolated
	 * @return Object for the given String data
	 * @throws IllegalAccessException,
	 *             IllegalArgumentException, InvocationTargetException, NoSuchMethodException, SecurityException, IllegalTermException
	 */
	public Object getObject(final String string) throws IllegalAccessException, IllegalArgumentException, InvocationTargetException, NoSuchMethodException,
			SecurityException, InstantiationException {

		final String predicate = getPredicate(string);

		if (predicate == null)
			return null;

		final Class<?> cl = getClass(predicate);

		// Not exist mapping between the predicate and the class
		if (cl == null)
			return null;

		final String[] parameters = getParameters(string);

		if (parameters == null)
			return null;

		final Object obj = cl.newInstance();

		populateObject(cl, parameters, obj);

		return obj;

	}

	/**
	 * @param string
	 *            The full value received as parameter of getObject
	 * @return All the Terms
	 */
	protected abstract String[] getParameters(final String string);

	/**
	 * @param string
	 *            The full value received as parameter of getObject
	 * @return The predicate name
	 * @throws IllegalArgumentException
	 */
	protected abstract String getPredicate(final String string) throws IllegalArgumentException;

	/**
	 * Returns data for the given Object
	 *
	 * @param obj
	 *            Object from witch data are extrapolated
	 * @return String data for the given Object in a String format
	 * @throws IllegalAccessException,
	 *             IllegalArgumentException, InvocationTargetException, NoSuchMethodException, SecurityException, IllegalTermException, IllegalTermException
	 */
	public String getString(final Object obj) throws IllegalAccessException, IllegalArgumentException, InvocationTargetException, NoSuchMethodException,
			SecurityException, ObjectNotValidException, IllegalAnnotationException, IllegalTermException {
		final String predicate = registerClass(obj.getClass());

		final HashMap<Integer, Object> parametersMap = new HashMap<>();
		for (final Field field : obj.getClass().getDeclaredFields()) {
			if (field.isSynthetic())
				continue;
			if (field.isAnnotationPresent(Param.class)) {
				final Object value = obj.getClass().getMethod("get" + Character.toUpperCase(field.getName().charAt(0)) + field.getName().substring(1))
						.invoke(obj);
				parametersMap.put(field.getAnnotation(Param.class).value(), value);
			}
		}
		return getActualString(predicate, parametersMap);
	}

	private void populateObject(final Class<?> cl, final String[] parameters, final Object obj) throws IllegalAccessException, InvocationTargetException {
		for (final Field field : cl.getDeclaredFields())
			if (field.isAnnotationPresent(Param.class)) {

				final int term = field.getAnnotation(Param.class).value();
				final String nameMethod = "set" + Character.toUpperCase(field.getName().charAt(0)) + field.getName().substring(1);
				final Method method = classSetterMethod.get(cl).get(nameMethod);

				if (method.getParameterTypes()[0].getName().equals(int.class.getName())
						|| method.getParameterTypes()[0].getName().equals(Integer.class.getName()))
					method.invoke(obj, Integer.valueOf(parameters[term]));
				else
					method.invoke(obj, parameters[term]);

			}
	}

	/**
	 * insert an object into {@link #predicateClass} and {@link #classSetterMethod}
	 *
	 * @return String representing pairing key of {@link #predicateClass}
	 */
	public String registerClass(final Class<?> cl) throws ObjectNotValidException, IllegalAnnotationException {

		final Annotation annotation = cl.getAnnotation(Id.class);

		if (annotation == null)
			throw new IllegalAnnotationException();

		final String predicate = ((Id) annotation).value();

		if (predicate.contains(" "))
			throw new ObjectNotValidException();

		// String predicate = cl.getAnnotation(Predicate.class).value();
		predicateClass.put(predicate, cl);
		final Map<String, Method> namesMethods = new HashMap<>();
		for (final Method method : cl.getMethods())
			if (method.getName().startsWith("set"))
				namesMethods.put(method.getName(), method);
		classSetterMethod.put(cl, namesMethods);
		return predicate;
	}

	public void unregisterClass(final Class<?> cl) throws IllegalAnnotationException {

		final Annotation annotation = cl.getAnnotation(Id.class);

		if (annotation == null)
			throw new IllegalAnnotationException();

		final String predicate = ((Id) annotation).value();

		predicateClass.remove(predicate);
		classSetterMethod.remove(cl);

	}

}

File: Python version/languages/mapper.py
from abc import ABCMeta, abstractmethod
from languages.predicate import Predicate
from languages.asp.symbolic_constant import SymbolicConstant


class Mapper(object):
    """Base class, contains methods used to transform Objects into
    InputProgram."""
    __metaclass__ = ABCMeta

    def __init__(self):
        self._objects = None
        self._parser = None
        # Represents a dict, where are stored a string name of a predicate as a
        # key and the corresponding Class element
        self._predicate_class = dict()

    @abstractmethod
    def _get_actual_string(self, predicate, parameters_map):
        pass

    @abstractmethod
    def _get_id(self):
        pass

    @abstractmethod
    def _get_param(self):
        pass

    def get_class(self, predicate):
        """Returns a string for the given predicate name string."""
        return self._predicate_class.get(predicate)

    @staticmethod
    def __populate_object(parameters, obj):
        """Sets a fields of object from set of parameters given, by invoking
        setters methods of the object."""
        for key, value in obj.get_terms_type().items():
            if isinstance(value, tuple) and len(value) == 2:
                name_method = "set_" + value[0]
                if value[1] is int:
                    getattr(obj, name_method)(int(parameters[key]))
                elif value[1] is SymbolicConstant:
                    getattr(obj, name_method)(
                        SymbolicConstant(parameters[key]))
            else:
                name_method = "set_" + value
                getattr(obj, name_method)(parameters[key])

    def get_object(self, atom):
        """Returns an Object for the given string.

        The parameter string is a string from which data are
        extrapolated. The method return a Object for the given string
        data.
        """
        predicate = self._get_id(atom)

        if predicate is None:
            return None

        cl = self.get_class(predicate)

        if cl is None:
            return None

        parameters = self._get_param(atom)

        if parameters is None:
            return None

        obj = cl()

        self.__populate_object(parameters, obj)

        return obj

    def register_class(self, cl):
        """Insert an object into _predicate_class.

        The method return a string representing pairing key of
        _predicate_class.
        """
        if not issubclass(cl, Predicate):
            raise "input class is not subclass of Predicate"
        predicate = cl.get_predicate_name()
        if " " in predicate:
            raise "Value of the object is not valid"
        self._predicate_class[predicate] = cl
        return predicate

    def unregister_class(self, cl):
        """Remove an object from _predicate_class."""
        if not issubclass(cl, Predicate):
            raise "input class is not subclass of Predicate"
        predicate = cl.get_predicate_name()
        del self._predicate_class[predicate]

    def get_string(self, obj):
        """Returns data for the given Object.

        The parameter obj is the Object from which data are
        extrapolated. The method return a string data for the given
        Object in a String format.
        """
        predicate = self.register_class(obj.__class__)

        parameters_map = dict()
        for key, value in obj.get_terms_type().items():
            if isinstance(value, tuple) and len(value) == 2:
                val = getattr(obj, "get_" + value[0])()
            else:
                val = getattr(obj, "get_" + value)()

            parameters_map[key] = val
        return self._get_actual_string(predicate, parameters_map)
